### DISCUSSION TITLE
Error: "IndexError: cannot do a non-empty take from an empty axes"

### DISCUSSION THREAD

**Discussion by Shakka1511** [jcreinhold/intensity-normalization#66] [Q&A] [answered]
Hello all,

Please excuse my relative inexperience with python/coding. 

I have been using the intensity-normalisation package to homogenise a dataset consisting of MRIs from brain tumors.
I have tried to use the nyul normalise to perform histogram matching on the dataset. Unfortunately, for some reason for about 1 in 20-25 images it gives the following index error:

> IndexError: cannot do a non-empty take from an empty axes

There does not seem to be a pattern or specific problem with these images.

Anyone else experienced the same problem or might have an idea what the issue could be?

**Comment by jcreinhold** [comment, marked as answer]
What is the file format of the images (e.g., NIfTI)? Are they 2D or 3D? Is there something different about the 1 image that is failing? Can you provide the full traceback (error message)?

**Reply by Shakka1511** [reply]
Thanks for the response!

Originally the images were .nrrd files which are 3D. However, the image that is being used is 2D as it is a single slice and has a shape of 256,256. There are no specific differences that I can see/find between the working/non-working images. However, I also don't understand the error completely so I have no idea where to look for such a difference. 


Complete error (see attachment):
[Error_Hist_match.txt](https://github.com/jcreinhold/intensity-normalization/files/9373855/Error_Hist_match.txt)

Thank you.

**Reply by jcreinhold** [reply]
Sorry for the delay. I'm not sure what's going on in the traceback because I'm missing context about the variables/functions you've defined. You can check the shape of all the images, perhaps an axis was dropped in the image that caused the failure (e.g., all the other images were (256, 256, 1) and the problematic image was (256, 256)).

However, the Nyul method only works on images of the same anatomy. If you split up the image volume into 2D slices, the slices will not be of the same anatomy. For example, in a head MR image, one axial slice will be of the neck and another will be of the brain. You need a set of images of the same anatomy (e.g., a set of brain MR images or a set of knee MR images) for the method to work correctly. See the original paper for details.

I wouldn't recommend Nyul as the first normalization scheme to try, however, because it suffers from the same problems of any histogram matching method; namely, if the distribution of the tissue classes is modestly different, the results will be terrible (e.g., in brain MR, you fit the method with healthy brain images and then use it to normalize the image of someone with enlarged ventricles). Even if you manually check all normalized images to make sure they qualitatively look reasonable, it's almost certainly not the right thing to do for reliable, trustworthy medical image analysis.
